From owner-linux-pci@atrey.karlin.mff.cuni.cz Wed Oct  3 11:14:41 2007
From: Roland Dreier <roland@digitalvampire.org>
Date: Wed, 03 Oct 2007 11:15:11 -0700
Subject: MSI: Use correct data offset for 32-bit MSI in read_msi_msg()
To: Greg KH <greg@kroah.com>
Cc: Eric W. Biederman <ebiederm@xmission.com>
Message-ID: <87abr0oyj4.fsf@shaolin.home.digitalvampire.org>


While reading the MSI code trying to find a reason why MSI wouldn't
work for devices that have a 32-bit MSI address capability, I noticed
that read_msi_msg() seems to read the message data from the wrong
offset in this case.

Signed-off-by: Roland Dreier <roland@digitalvampire.org>
Acked-by: Eric W. Biederman <ebiederm@xmission.com>
Cc: stable <stable@kernel.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>


---
 drivers/pci/msi.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/pci/msi.c
+++ b/drivers/pci/msi.c
@@ -132,7 +132,7 @@ void read_msi_msg(unsigned int irq, stru
 			pci_read_config_word(dev, msi_data_reg(pos, 1), &data);
 		} else {
 			msg->address_hi = 0;
-			pci_read_config_word(dev, msi_data_reg(pos, 1), &data);
+			pci_read_config_word(dev, msi_data_reg(pos, 0), &data);
 		}
 		msg->data = data;
 		break;
